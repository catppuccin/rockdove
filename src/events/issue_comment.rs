use octocrab::models::webhook_events::{
    payload::{IssueCommentWebhookEventAction, IssueCommentWebhookEventPayload},
    WebhookEvent,
};

use crate::{
    colors::{ISSUE_COLOR, PULL_REQUEST_COLOR},
    embed_builder::EmbedBuilder,
};

pub fn make_issue_comment_embed(
    event: WebhookEvent,
    specifics: &IssueCommentWebhookEventPayload,
) -> Option<EmbedBuilder> {
    if !matches!(specifics.action, IssueCommentWebhookEventAction::Created) {
        return None;
    }

    let repo = event
        .repository
        .expect("issue comment events should always have a repository");

    let mut embed = EmbedBuilder::default();

    let repo_name = repo.full_name.unwrap_or(repo.name);
    let target = if specifics.issue.pull_request.is_some() {
        "pull request"
    } else {
        "issue"
    };

    embed.title(&format!(
        "[{}] New comment on {} #{}: {}",
        repo_name, target, specifics.issue.number, specifics.issue.title,
    ));

    embed.url(specifics.comment.html_url.as_str());

    if let Some(ref body) = specifics.comment.body {
        embed.description(body);
    }

    embed.color(if specifics.issue.pull_request.is_some() {
        PULL_REQUEST_COLOR
    } else {
        ISSUE_COLOR
    });

    Some(embed)
}

#[cfg(test)]
mod tests {
    use crate::{
        make_embed,
        tests::{embed_context, TestConfig},
    };
    use std::fs;
    use yare::parameterized;

    #[parameterized(
        created = { "created" },
        created_on_pull_request = { "created_on_pull_request" },
      )]
    fn snapshot(event_type: &str) {
        let event = "issue_comment";
        let root = env!("CARGO_MANIFEST_DIR");
        let filename = format!("{root}/fixtures/{event}/{event_type}.json");
        let payload = fs::read_to_string(&filename).expect("fixture exists");
        let TestConfig {
            webhook_event,
            mut settings,
        } = TestConfig::new(event, &payload);

        let embed = make_embed(webhook_event)
            .expect("make_embed should succeed")
            .expect("event fixture can be turned into an embed");

        settings.set_info(&embed_context(&embed));
        settings.bind(|| insta::assert_yaml_snapshot!(embed));
    }
}
