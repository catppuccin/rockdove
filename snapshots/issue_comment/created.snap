---
source: src/events/issue_comment.rs
info:
  author_name_length: 8
  colour_hex: "#A6E3A1"
  description_length: 7
  title_length: 75
---
embeds:
  - author:
      icon_url: "https://avatars.githubusercontent.com/u/58985301?v=4"
      name: sgoudham
      url: "https://github.com/sgoudham"
    color: 10937249
    description: Testing
    title: "[catppuccin/java] New comment on issue #20: Reconsider OSSRH Authentication"
    url: "https://github.com/catppuccin/java/issues/20#issuecomment-2351090061"
